chalice/package.py

```diff
@@ -530,15 +559,272 @@ def _register_cfn_resource_name(self, name):
         return cfn_name
 
 
+class TerraformGenerator(TemplateGenerator):
+
+    template_file = "chalice.tf.json"
+
+    def generate(self, resources):
+        # type: (List[models.Model]) -> Dict[str, Any]
+        template = {
+            'resource': {},
+            'terraform': {
+                'required_version': '> 0.11.0'
+            },
+            'provider': {
+                'aws': {'version': '> 2.0.0'},
+            },
+            'data': {
+                'aws_caller_identity': {'chalice': {}},
+                'aws_region': {'chalice': {}}
+            }
+        }
+        for resource in resources:
+            self.dispatch(resource, template)
+        return template
+
+    def _fref(self, lambda_function, attr='arn'):
+        # type: (models.ManagedModel, str) -> str
+        return '${aws_lambda_function.%s.%s}' % (
+            lambda_function.resource_name, attr)
+
+    def _arnref(self, arn_template, **kw):
+        # type: (str, str) -> str
+        d = dict(
+            region='${data.aws_region.chalice.name}',
+            account_id='${data.aws_caller_identity.chalice.account_id}')
+        d.update(kw)
+        return arn_template % d
+
+    def _generate_managediamrole(self, resource, template):
+        # type: (models.ManagedIAMRole, Dict[str, Any]) -> None
+        template['resource'].setdefault('aws_iam_role', {})[
+            resource.resource_name] = {
```

Comment:
Just making a general comment for future development. I went through the resource names we are generating. They look fine too me. The one thing that we should address (but do not need to now) is add tests to make sure that we do not accidentally change `resource_name`s and have that cause a change in the resource name in the generated template. This is especially true for terraform as people will be able to reference resources from the generated chalice template to other templates. The backwards compatibility guarantee will be needed for this PR as well: https://github.com/aws/chalice/pull/1195